core/src/main/java/org/elasticsearch/common/xcontent/XContentParserUtils.java

```diff
@@ -121,23 +121,26 @@ public static Object parseStoredFieldsValue(XContentParser parser) throws IOExce
      * @param parser      the current {@link XContentParser}
      * @param delimiter   the delimiter to use to splits the field's name
      * @param objectClass the object class of the object to parse
+     * @param consumer    something to consume the parsed object
      * @param <T>         the type of the object to parse
-     * @return the parsed object
      * @throws IOException if anything went wrong during parsing or if the type or name cannot be derived
      *                     from the field's name
      */
-    public static <T> T parseTypedKeysObject(XContentParser parser, String delimiter, Class<T> objectClass) throws IOException {
+    public static <T> void parseTypedKeysObject(XContentParser parser, String delimiter, Class<T> objectClass, Consumer<T> consumer)
+            throws IOException {
+        assert parser.currentToken() == XContentParser.Token.START_OBJECT || parser.currentToken() == XContentParser.Token.START_ARRAY;
         String currentFieldName = parser.currentName();
         if (Strings.hasLength(currentFieldName)) {
             int position = currentFieldName.indexOf(delimiter);
             if (position > 0) {
                 String type = currentFieldName.substring(0, position);
                 String name = currentFieldName.substring(position + 1);
-                return parser.namedObject(objectClass, type, name);
+                consumer.accept(parser.namedObject(objectClass, type, name));
+                return;
             }
         }
-        throw new ParsingException(parser.getTokenLocation(), "Cannot parse object of class [" + objectClass.getSimpleName()
-                + "] without type information. Set [" + RestSearchAction.TYPED_KEYS_PARAM + "] parameter on the request to ensure the"
-                + " type information is added to the response output");
+        // If the field name is empty or we didn't find a delimiter we ignore
+        // the object for forward compatibility instead of throwing an error
```

Comment:
Yes, thats the assumption. I think that should rather go into the method description though, its something we expect the caller of this method to make sure, so better have it in the method docs.